projects/event-lambdas/src/redrive-from-S3-lambda/index.ts

```diff
@@ -0,0 +1,107 @@
+import {getEventsFromS3File, putEventsToKinesisStream} from "../lib/util";
+import {s3} from "../lib/aws";
+import {telemetryBucketName} from "../lib/constants";
+import {Marker} from "aws-sdk/clients/s3";
+
+const oneMinInMillis = 60 * 1000;
+const tenMinsInMillis = 10 * oneMinInMillis;
+const fourteenMinsInMillis = 14 * oneMinInMillis;
+
+type MarkerDone = null;
+
+export const handler = async (
+  event: { Payload?: Marker | unknown }
+): Promise<Marker | MarkerDone> => {
+
+  const startTimeEpoch = new Date().getTime();
+
+  if(typeof event?.Payload !== "undefined" && typeof event?.Payload !== "string") {
+    console.error("Invalid input.", event)
+    throw Error("Invalid input. Expected string, got " + typeof event.Payload);
+  }
+
+  const maybeStartingMarker = event?.Payload;
+
+  const processPageOfFiles = async (maybeMarker: Marker | undefined): Promise<Marker | MarkerDone> => {
+
+    // we could instead turn on bucket inventory, and work through that
+    // (this would mean we could start with latest, working back through time, as more recent is probably most useful/urgent)
+    const pageOfFilesResponse = await s3.listObjects({
+      Bucket: telemetryBucketName,
+      Marker: maybeMarker,
+      MaxKeys: 500, // max allowed
+    }).promise();
+
+    const pageOfFiles = pageOfFilesResponse.Contents;
+
+    if(!pageOfFiles || pageOfFiles.length === 0) {
+      console.log("No files to process")
+      return null;
+    }
+
+
+    console.log(`page of files: ${pageOfFiles.length}, starting at ${pageOfFilesResponse.Marker}`);
+
+    const eventsArrays = await Promise.all(pageOfFiles.map(async (file) => {
+      const Key = file.Key;
+
+      if(!Key) {
+        console.error("No key found in file", file);
+        return [];
+      }
+
+      const position = pageOfFiles.indexOf(file) + 1;
+
+      console.log(`Attempting to read from file (${position} of ${pageOfFiles.length}) at s3://${telemetryBucketName}/${Key}`);
+
+      const maybeEvents = await getEventsFromS3File(Key);
+
+      if (maybeEvents.error) {
+        console.error(`Invalid data in file with key ${Key}`, maybeEvents.error);
+        return [];
+      }
+
+      return maybeEvents.value;
+    }));
+
+    const events = eventsArrays.flat();
+
+    const chunkSize = 50;
```

Comment:
before this comment, I'd already been thinking how crude my chunking was, so had already been thinking of improvements, which led to https://github.com/guardian/editorial-tools-user-telemetry-service/pull/44/commits/d9475f001ad5cab3c63307231ae400e0661d5b57

Perhaps even better than the step function manipulating the shared kinesis stream (its a single stream for CODE and PROD and shared with the event-s3-lambda which is sending events as they happen) would be to change the provisioning mode to on-demand 🤔

Reply:
> before this comment, I'd already been thinking how crude my chunking was, so had already been thinking of improvements, which led to [d9475f0](https://github.com/guardian/editorial-tools-user-telemetry-service/commit/d9475f001ad5cab3c63307231ae400e0661d5b57) - which optimises for 1MB/sec explicitly

so far this is resulting in around 20k records per min - reasonable, but with on-demand shards we could go much faster (although wonder at what rate ELK could consume)

Reply:
_> provisioned to on-demand 🤔_

This seems sensible. Guess we might drop some events on the leading edge of the scale-out event — unsure how quickly that happens, may be an acceptable risk. We won't be redriving often.

The above might be a reason to include this in the step fn — scale out, then redrive, to ensure we're not at risk of saturating the stream at the expense of actual telemetry.

Reply:
Looks like the `telemetry-stream-PROD` Cloudformation stack was created using https://github.com/guardian/deploy-tools-platform/blob/main/cloudformation/kinesis-logging-stack.yaml (which looks to be _shared_, I'm tempted to add an optional param to set the capacity mode)

Reply:
> > Looks like the `telemetry-stream-PROD` Cloudformation stack was created using [guardian/deploy-tools-platform@`main`/cloudformation/kinesis-logging-stack.yaml](https://github.com/guardian/deploy-tools-platform/blob/main/cloudformation/kinesis-logging-stack.yaml) (which looks to be _shared_), so I'm tempted to add an optional param to set the capacity mode
> 
> [guardian/deploy-tools-platform#842](https://github.com/guardian/deploy-tools-platform/pull/842)

Looks like `telemetry-stream-PROD` is quite behind vs https://github.com/guardian/deploy-tools-platform/blob/main/cloudformation/kinesis-logging-stack.yaml, notably including encryption (which I don't think is necessary for this stream, but is hard coded in the template these days, not sure I fancy making that parameterised though as the 'default' will be problematic), I'm going to download the existing template for https://github.com/guardian/deploy-tools-platform/blob/main/cloudformation/kinesis-logging-stack.yaml and modify it just for the capacity mode.

Reply:
in https://github.com/guardian/editorial-tools-user-telemetry-service/pull/44/commits/a8a78aaf0107c5c3fd4d8e70b531b850bf1c69bc I tried going full blast after making the stream ON_DEMAND, but we see major throttling, the on_demand doesn't scale fast enough, I will indeed need to scale explicitly within the lambda 😞

Reply:
https://github.com/guardian/editorial-tools-user-telemetry-service/pull/44/commits/f7018cf81c5b428c4319eba5ad030dadd6c1ec25 adds our own _on-demand scaling_ (up to 32 shards) - since ON_DEMAND capacity mode doesn't scale immediately we still get throttled, so we've switched back to provisioned and we do our own gradual scale up

Reply:
> [4b564e7](https://github.com/guardian/editorial-tools-user-telemetry-service/commit/4b564e7feb6c6ad44b46a3f6572122acae61214a) adds our own _on-demand scaling_ (up to 32 shards) - since ON_DEMAND capacity mode doesn't scale immediately we still get throttled, so we've switched back to provisioned and we do our own gradual scale up

see `Testing` section of description to see how this performed, we expect PROD to take 12 hours, reasonable I feel